common/eth2/src/types.rs

```diff
@@ -1008,15 +1010,42 @@ impl FromStr for Accept {
     type Err = String;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match s {
-            "application/octet-stream" => Ok(Accept::Ssz),
-            "application/json" => Ok(Accept::Json),
-            "*/*" => Ok(Accept::Any),
-            _ => Err("accept header cannot be parsed.".to_string()),
-        }
+        let mimes = parse_accept(s);
+
+        // [q-factor weighting]: https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.2
+        // find the highest q-factor supported accept type
+        mimes.and_then(|mut ms| {
+            ms.sort_by(|m1, m2| {
```

Comment:
You can make this even shorter if you use [`sort_by_key`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.sort_by_key), and we can avoid duplicating the `.get_param("q").map_or(..)` stuff.